### PR TITLE
Feat: Add spec and parser for 'crio.conf'

### PR DESCRIPTION
**Pull request by shlao**
Signed-off-by: shlao <shlao@redhat.com>

### All Pull Requests:

Check all that apply:

* [x] Have you followed the guidelines in our Contributing document, including the instructions about commit messages?
* [ ] Is this PR to correct an issue?
* [x] Is this PR an enhancement?

### Complete Description of Additions/Changes:

The OpenShift Node will fail to start the cri-o service because the `conmon` sets improperly  in 'crio.conf' or '/etc/crio/crio.conf.d/*-conmon.conf'. Detail info please see: INSIGHTOCP-525 card.

### CODE REVIEW

**Comment by clarabez** [comment]
Hi, @shlao  !
We've identified that setup pipeline is failing for RHEL6 on "QE Pull Request Testing", but it is working as expected for RHEL79, RHEL84 and RHEL90.

This issue has been reported and is being analyzed here: [RHCLOUD-17392](https://issues.redhat.com/browse/RHCLOUD-17392).

**Comment by shlao** [comment]
@clarabez  Thanks for the info.

**Comment by shlao** [comment]
Waiting the [PR-3310](https://github.com/RedHatInsights/insights-core/pull/3310)

**Comment by shlao** [comment]
@xiangce Will you review the updated PR? Thanks!

File: insights/combiners/crio_conf.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+            for section in src.sections():
+                if section not in dest:
+                    dest[section] = {}
+
+                for option in src.items(section):
+                    dest[section][option] = src.get(section, option)
+
+        for crio_conf in crio_confs:
+            self.files.append(crio_conf.file_path)
+            if crio_conf.file_name == "crio.conf":
```
Comment (xiangce) [inline insights/combiners/crio_conf.py:92]: Would it be better to use the full path in this comparison?

File: insights/combiners/crio_conf.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                main_data = crio_conf
+            else:
+                dict_merge(self.data, crio_conf)
+
+        dict_merge(self.data, main_data)
```
Comment (xiangce) [inline insights/combiners/crio_conf.py:97]: Per the manual of the `crio.conf.d`, this is incorrect, the configurations in `crio.conf.d` will overwrite the configuration in the default `crio.conf`.
```
When it exists, the main configuration file (/etc/crio/crio.conf by default)
is read before any file in the configuration directory (/etc/crio/crio.conf.d). 
Settings in that file have the lowest precedence.
```

https://man.archlinux.org/man/community/cri-o/crio.conf.d.5.en

File: insights/specs/sos_archive.py [hunk 3]
```diff
@@ -52,6 +52,7 @@ class SosSpecs(Specs):
     cpu_vulns_spec_store_bypass = simple_file("sys/devices/system/cpu/vulnerabilities/spec_store_bypass")
     cpuinfo_max_freq = simple_file("/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq")
     cpupower_frequency_info = simple_file("sos_commands/processor/cpupower_frequency-info")
+    crio_conf = glob_file([r"etc/crio/crio.conf", r"etc/crio/crio.conf.d/*-conmon.conf"])
```
Comment (xiangce) [inline insights/specs/sos_archive.py:55]: Why here only read the `*-conmon.conf` from the `/etc/crio/crio.conf.d/` dir? If I understand correctly, you should read all the `.conf` file in this dir.

**Review by xiangce — CHANGES_REQUESTED** [review]

File: insights/parsr/iniparser.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import string
 
-from insights.parsr import (EOF, HangingString, InSet, LeftBracket, Lift,
+from insights.parsr import (Comma, EOF, EOL, DoubleQuotedString, HangingString, InSet, LeftBracket, Lift,
```
Comment (xiangce) [inline insights/parsr/iniparser.py:3]: @shlao - would you please adjust the length of these lines to make it a bit alignment and not exceed the 75? Thanks.

**Comment by shlao** [comment]
@xiangce  Will you help review.

File: insights/combiners/crio_conf.py (hunk 1 above)
Comment (shlao) [reply insights/combiners/crio_conf.py]: I fxied it. Will you help me confirm it?

File: insights/combiners/crio_conf.py (hunk 2 above)
Comment (shlao) [reply insights/combiners/crio_conf.py:100]: That's right. Will you help me confirm it?

File: insights/specs/sos_archive.py (hunk 3 above)
Comment (shlao) [reply insights/specs/sos_archive.py]: I updated it to "etc/crio/crio.conf.d/*-conmon.conf", will you help me review it.

File: insights/combiners/crio_conf.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                if section not in dest:
+                    dest[section] = {}
+
+                for option in src.items(section):
+                    dest[section][option] = src.get(section, option)
+
+        for crio_conf in crio_confs:
+            self.files.append(crio_conf.file_path)
+            if crio_conf.file_path == "/etc/crio/crio.conf":
+                main_data = crio_conf
```
Comment (xiangce) [inline insights/combiners/crio_conf.py:96]: nitpick: the variable `main_data` is not required, you can merge it to the `self.data` here directly.

File: insights/combiners/crio_conf.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            else:
+                conf_d_data.append(crio_conf)
+
+        dict_merge(self.data, main_data)
+        for crio_conf in conf_d_data:
```
Comment (xiangce) [inline insights/combiners/crio_conf.py:101]: I believe you should sort the `confs` got from `crio.conf.d` with the `file_name` at first before the merge.

```
Files in the configuration directory are sorted by name in lexical order and applied in that order. If multiple configuration files specify the same configuration option the setting specified in file sorted last takes precedence over any other value. That is if both 00-default.conf and 10-custom.conf exist in crio.conf.d and both specify different values for a certain configuration option the value from 10-custom.conf will be applied.
```
https://man.archlinux.org/man/community/cri-o/crio.conf.d.5.en

**Review by xiangce — CHANGES_REQUESTED** [review]

File: insights/combiners/crio_conf.py (hunk 5 above)
Comment (shlao) [reply insights/combiners/crio_conf.py]: I removed the redundant main_data. Please have a look.

File: insights/combiners/crio_conf.py (hunk 6 above)
Comment (shlao) [reply insights/combiners/crio_conf.py:107]: Sorted the data with file_name. Please have a look.

**Review by xiangce — APPROVED** [review]

**Comment by xiangce** [comment]
This PR modified the base `parse_doc` function to support the value of a key which is a list in multiple lines, @bfahr would you please also have a look to the changes?

**Comment by shlao** [comment]
@bfahr   Will you help me confirm the enhancement for iniparser. There is rule depends on it

The enhancement supports the following options in INI format:
-  Formation No.1
```
plugin_dirs = [
] 
```
-  Formation No.2
```
plugin_dirs = [
    "/usr/libexec/cni",
] 
```
- Formation No.3
```
plugin_dirs = [
    "/usr/libexec/cni",
    "/etc/network/cni",
] 
```

**Comment by shlao** [comment]
The enhancement supports the following options in INI:
-  Formation No.4    (White space, `\t`)
```
plugin_dirs = [
  \t
] 
```

-  Formation No.5    (No Comma)
```
plugin_dirs = [
    "/usr/libexec/cni"
] 
```

-  Formation No.6    (Comment item)
```
plugin_dirs = [
    "/usr/libexec/cni",
    # Some Comment
] 
```

**Review by bfahr — APPROVED** [review]
Changes look good to me.

**Pull request merged**